### PR TITLE
Adding additional BC validation and support for reflecting BCs.

### LINKED ISSUE #31: Code not reporting an error for incorrect flow_condition
The code is incorrectly running without an error for the following `.yaml` in which the `momentum` is commented out for one of the `flow_conditions`:

```
# This test input corresponds to toy-problems/swe_roe/ex2b.c.

physics:
  flow:
    mode: swe

numerics:
  spatial: fv
  temporal: euler
  riemann: roe

logging:
  level: debug

time:
  final_time: 0.005
  unit: hours
  max_step: 1 #1000 

grid:
  file: planar_dam_10x5.msh

initial_conditions:
  1: # upstream region
    flow: dam_top_ic
  2: # downstream region
    flow: dam_bottom_ic

# NOTE: We use reflecting boundary conditions for the boundary, top_wall, and
# NOTE: bottom_wall surfaces

flow_conditions:
  dam_top_ic:
    type: dirichlet
    height: 10
    momentum: [0, 0]
  dam_bottom_ic:
    type: dirichlet
    height: 5
    #momentum: [0, 0]
```
NOTE: The `max_step` is modified to be `1`.

You can run the code via:
```
cd build/driver/tests/swe_roe
../../rdycore ex2b.yaml -dm_view -ts_monitor

DETAIL: ==========================================================
DETAIL: RDycore (input read from ex2b.yaml)
DETAIL: ==========================================================
DETAIL: Physics:
DETAIL:   Sediment model: disabled
DETAIL:   Salinity model: disabled
DETAIL:   Bed friction model: disabled
DETAIL: Numerics:
DETAIL:   Spatial discretization: finite volume (FV)
DETAIL:   Temporal discretization: forward euler
DETAIL:   Riemann solver: roe
DETAIL: Time:
DETAIL:   Final time: 0.005 hours
DETAIL: Logging:
DETAIL:   Primary log file: <stdout>
DETAIL: Restart:
DETAIL:   (disabled)
DEBUG: Creating DMs...
DM Object: grid 1 MPI process
  type: plex
grid in 2 dimensions:
  Number of 0-cells per rank: 63
  Number of 1-cells per rank: 106
  Number of 2-cells per rank: 44
Labels:
  celltype: 3 strata with value/size (0 (63), 4 (44), 1 (106))
  depth: 3 strata with value/size (0 (63), 1 (106), 2 (44))
  Cell Sets: 2 strata with value/size (1 (20), 2 (24))
  Face Sets: 3 strata with value/size (1 (26), 2 (4), 3 (6))
  boundary_edges: 1 strata with value/size (1 (36))
Field height:
  adjacency FVM++
Field x momentum:
  adjacency FEM
Field y momentum:
  adjacency FEM
DEBUG: Initializing regions and boundaries...
DEBUG:   Found region 1 (20 cells)
DEBUG:   Found region 2 (24 cells)
DEBUG:   Found boundary 1 (26 edges)
DEBUG:   Found boundary 2 (4 edges)
DEBUG:   Found boundary 3 (6 edges)
DEBUG: Initializing initial/boundary conditions and sources...
DEBUG: Creating solvers and vectors...
DEBUG: Creating FV mesh...
DEBUG: Initializing solution data...
DEBUG: Creating output directory output...
DEBUG: Running simulation...
0 TS dt 18. time 0.
DEBUG: Max courant number 178.245 encountered at edge 58 of cell 33 is 178.245449
1 TS dt 18. time 18.
DEBUG: Writing simulation output...
```

### LINKED ISSUE #32: Support CONDITION_REFLECTING
I believe the `CONDITION_REFLECTING` in `ParseFlowConditions()` can be supported via:

```
git diff src/read_config_file.c
diff --git a/src/read_config_file.c b/src/read_config_file.c
index 7147084..da6bee6 100644
--- a/src/read_config_file.c
+++ b/src/read_config_file.c
@@ -694,7 +694,7 @@ static PetscErrorCode ParseFlowConditions(yaml_event_t *event, YamlParserState *
       } else {
         if (!strcmp(state->parameter, "type")) {
           PetscInt selection;
-          SelectItem(value, 2, (const char *[2]){"dirichlet", "neumann"}, (PetscInt[2]){CONDITION_DIRICHLET, CONDITION_NEUMANN}, &selection);
+          SelectItem(value, 3, (const char *[3]){"dirichlet", "neumann", "reflecting"}, (PetscInt[3]){CONDITION_DIRICHLET, CONDITION_NEUMANN, CONDITION_REFLECTING}, &selection);
           PetscCheck(selection != -1, state->comm, PETSC_ERR_USER, "Invalid flow condition %s.type: %s", flow_cond->name, value);
           flow_cond->type     = selection;
           state->parameter[0] = 0;  // clear parameter name
```

After this change, a new `flow_condition` can be added via:

```
flow_conditions:
  wall_condition:
    type: reflecting
    height: ???
    momentum: ???
```

but, what should be the entries for `height` and `momentum` because these entries will not be used and automatically created by the code but computed on the fly? Should there be a check that if `type == reflecting`, there should be no entries for `height` and `momentum`?

### PR DESCRIPTION
**Pull request by jeff-cohere**
Closes #31
Closes #32

### CODE REVIEW

**Comment by bishtgautam** [comment]
How about also editing the [ex2b.yaml](https://github.com/RDycore/RDycore/blob/main/driver/tests/swe_roe/ex2b.yaml) to add boundary conditions of the `reflecting` type for one/two boundaries out of the three boundaries in [planar_dam_10x5.msh#L6](https://github.com/RDycore/RDycore/blob/main/share/meshes/planar_dam_10x5.msh#L6)? The solution should remain unchanged after you add the new BCs because by default `reflecting` BC is automatically applied internally.

**Review by bishtgautam — APPROVED** [review]
